ast/src/parsed/display.rs

```diff
@@ -549,7 +595,13 @@ impl<Ref: Display> Display for Expression<Ref> {
             Expression::Tuple(items) => write!(f, "({})", format_list(items)),
             Expression::LambdaExpression(lambda) => write!(f, "{}", lambda),
             Expression::ArrayLiteral(array) => write!(f, "{array}"),
-            Expression::BinaryOperation(left, op, right) => write!(f, "({left} {op} {right})"),
+            Expression::BinaryOperation(left, op, right) => write!(
+                f,
+                "{} {} {}",
+                format_expression_with_precedence(left, self.binary_op_precedence()),
```

Comment:
Hm, we could make this a member function of Expression and call it `format_with_precedence`, then this monster would get a bit shorter.